### PR TITLE
Move history API logic into its own package

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Move start workflow validation into api package
* Move signal workflow validation into api package

**Why?**
Follow up of #2778: Code simplification & restructure

**How did you test it?**
N/A

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

File: service/history/api/metrics_scope.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	"context"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/rpc/interceptor"
+)
+
+func MetricsScope(
```
Comment (yux0) [inline service/history/api/metrics_scope.go:35]: should it be in common package?

File: service/history/api/update_workflow_util.go [hunk 2]
```diff
@@ -88,7 +88,11 @@ UpdateHistoryLoop:
 			if err != nil {
 				return err
 			}
-			if err = NewWorkflowVersionCheck(shard, mutableState, newMutableState); err != nil {
+			lastWriteVersion, err := mutableState.GetLastWriteVersion()
```
Comment (yux0) [inline service/history/api/update_workflow_util.go:91]: nit: what about update_workflow.go and signal_workflow.go

Comment (wxing1292) [reply service/history/api/update_workflow_util.go:91]: i plan to use e.g. `signal_workflow.go` for signal API, `*_util.go` for common utility

or do you have a proposal for naming convention?

Comment (wxing1292) [reply service/history/api/update_workflow_util.go:91]: current situation: signal & start & signal with start API share lots of common logic, but each API choose to impl the logic differently. so i plan to put all commonly used building blocks in `*_util.go`

Comment (yux0) [reply service/history/api/update_workflow_util.go:91]: misunderstand your intention. I am fine with the naming

**Review by yycptt — APPROVED** [review]

**Review by yux0 — APPROVED** [review]

**Pull request merged**